tikv/raftstore/snap.go

```diff
@@ -0,0 +1,834 @@
+package raftstore
+
+import (
+	"bytes"
+	"fmt"
+	"hash"
+	"hash/crc32"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/coocood/badger"
+	"github.com/coreos/etcd/raft/raftpb"
+	"github.com/ngaut/log"
+	"github.com/ngaut/unistore/rocksdb"
+	"github.com/ngaut/unistore/util"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/kvproto/pkg/metapb"
+	rspb "github.com/pingcap/kvproto/pkg/raft_serverpb"
+)
+
+type CFName = string
+
+const (
+	CFDefault CFName = "default"
+	CFLock    CFName = "lock"
+	CFWrite   CFName = "write"
+	CFRaft    CFName = "raft"
+
+	snapGenPrefix       = "gen" // Name prefix for the self-generated snapshot file.
+	snapRevPrefix       = "rev" // Name prefix for the received snapshot file.
+	sstFileSuffix       = ".sst"
+	tmpFileSuffix       = ".tmp"
+	cloneFileSuffix     = ".clone"
+	metaFileSuffix      = ".meta"
+	snapshotVersion     = 2
+	deleteRetryMaxTime  = 6
+	deleteRetryDuration = 500 * time.Millisecond
+
+	JobStatusPending    uint64 = 0
+	JobStatusRunning    uint64 = 1
+	JobStatusCancelling uint64 = 2
+	JobStatusCancelled  uint64 = 3
+	JobStatusFinished   uint64 = 4
+	JobStatusFailed     uint64 = 5
+)
+
+var (
+	snapshotCFs  = []CFName{CFDefault, CFLock, CFWrite}
+	defaultCFIdx = 0
+	lockCFIdx    = 1
+	writeCFIdx   = 2
+
+	errAbort = errors.New("abort")
+)
+
+type SnapKey struct {
+	RegionID uint64
+	Term     uint64
+	Index    uint64
+}
+
+func (k *SnapKey) String() string {
+	return fmt.Sprintf("%d_%d_%d", k.RegionID, k.Term, k.Index)
+}
+
+func SnapKeyFromRegionSnap(regionID uint64, snap *raftpb.Snapshot) *SnapKey {
+	return &SnapKey{
+		RegionID: regionID,
+		Term:     snap.Metadata.Term,
+		Index:    snap.Metadata.Index,
+	}
+}
+
+func SnapKeyFromSnap(snap *raftpb.Snapshot) (*SnapKey, error) {
+	data := new(rspb.RaftSnapshotData)
+	err := data.Unmarshal(snap.Data)
+	if err != nil {
+		return nil, err
+	}
+	return &SnapKey{
+		RegionID: data.Region.Id,
+		Term:     snap.Metadata.Term,
+		Index:    snap.Metadata.Index,
+	}, nil
+}
+
+type SnapStatistics struct {
+	Size    uint64
+	KVCount int
+}
+
+type ApplyOptions struct {
+	DBSnap    *DBSnapshot
```

Comment:
In TiKV, this field is RocksDB::DB, DBSnapshot is a struct combine badger and memstore, i prefer rename DBSnapshot to UniDB.